src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java

```diff
@@ -47,6 +50,20 @@ public KafkaProducerMessageHandler(final KafkaProducerContext kafkaProducerConte
 		this.kafkaProducerContext = kafkaProducerContext;
 	}
 
+	/**
+	 * Enables the use of headers for determining the target topic and partition of outbound messages. By default it is
+	 * set to true, but it can be disabled when those values are produced by upstream components that read messages
+	 * from Kafka sources themselves.
+	 *
+	 * @see KafkaHeaders#TOPIC
+	 * @see KafkaHeaders#PARTITION_ID
+	 * @param enableHeaderRouting whether the topic and destination headers should be considered
```

Comment:
This would be possible too - I am just concerned that disabling headers altogether would be too much of a breaking change. I think that this strikes a pretty fair balance, also given that the `<header-filter>` solution may inadvertently remove headers that are not used in routing, e.g. `kafka_messageKey`.